lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -0,0 +1,580 @@
+                if (child.getOperandNumber() == 0)
+                  return (Value) nullptr;
+                // Get the aggregate base type Value, if the dest is a subfield
+                // of it. Returns an invalid FieldRef, if this Value can be
+                // ignored.
+                auto fRef = getFieldRef(connect.getDest());
+
+                if (fRef.getValue() == nullptr)
+                  return (Value) nullptr;
+                if (inputArg.isValid())
+                  if (auto sinkOut = fRef.getValue().dyn_cast<BlockArgument>())
+                    // Record the combinational path that was discovered between
+                    // the ports of this module.
+                    portPaths[inputArg.val][inputArg.fieldId].push_back(
+                        Node(sinkOut, fRef.getFieldID()));
+
+                return connect.getDest();
+              })
+              .Case<RegOp, RegResetOp, SubaccessOp>(
+                  [&](auto reg) { return (Value) nullptr; })
+              .Default([&](Operation *op) {
+                if (op->getNumResults() == 1) {
+                  return (Value)op->getResult(0);
+                }
+                return (Value) nullptr;
+              });
+      if (childVal != nullptr) {
+        children.push_back({childVal, owner});
+      }
+    }
+  }
+
+  /// Record all the leaf ground type values for an aggregate type value. The
+  /// Optional `fieldIdsToAdd` can be used to only add field ids that exist in
+  /// the set and ignore all other fields. This is used for MemOp ports.
+  void gatherAggregateLeafs(Value val,
+                            llvm::SmallDenseSet<size_t> fieldIdsToAdd = {}) {
+    auto valType = val.getType().dyn_cast<FIRRTLBaseType>();
+    // Ignore RefType and GroundType Values.
+    if (!valType || valType.isGround())
+      return;
+    SmallVector<Node> worklist;
+    worklist.push_back(Node(val, 0));
+    // Assign a unique storage location for the `val`, which is the aggregate
+    // base Value.
+    auto baseStorageLocId = storageLocId;
+    valToStorageLocIdMap[val] = {baseStorageLocId};
+    storageLocIdToValuesMap[baseStorageLocId] = {val};
+    // Reserve unique storage location ids for all the subfields of `val`.
+    storageLocId += valType.getMaxFieldID() + 1;
+    auto handleSub = [&](Value res, size_t fieldId, bool isGround) {
+      if (isGround) {
+        if (!fieldIdsToAdd.empty())
+          if (!fieldIdsToAdd.contains(fieldId))
+            return;
+        // Compute the storage location id based on the fieldIndex.
+        auto storageLocId = baseStorageLocId + fieldId;
+        // The results of subaccess op can refer to multiple storage locations.
+        valToStorageLocIdMap[res].push_back(storageLocId);
+        storageLocIdToValuesMap[storageLocId].push_back(res);
+        leafToBaseValMap[res] = val;
+      } else
+        worklist.push_back(Node(res, fieldId));
```

Comment:
nit: Is `isGround == res.getType().isGround()` always true? If so, we can remove `isGround` argument and instead we can use `res.getType().isGround()`.

Reply:
Right, removed the argument.